File: Version/方案1/build_lib.py
import os
import sys
import subprocess
from pathlib import Path

# ================================
# 构建策略开关
# 0 = 保留 version.c / version.h
# 1 = 生成 lib 后删除
# ================================
CLEAN_VERSION_SRC = 0

# ================================
# Keil 安装路径
# ================================
KEIL_PATH = Path(r"E:\Keil")

def run(cmd):
    subprocess.check_call(cmd)

def must_exist(path: Path, name: str):
    if not path.exists():
        print(f"[ERROR] {name} not found: {path}")
        sys.exit(1)

# ================================
# 获取程序真实所在目录
# py 运行：__file__
# exe 运行：sys.executable
# ================================
def get_app_dir() -> Path:
    if getattr(sys, "frozen", False):
        return Path(sys.executable).resolve().parent
    else:
        return Path(__file__).resolve().parent

def main():
    print("[INFO] Build GD32 FMC + Version static library (ARMCC5)")

    # ================================
    # Version 目录（脚本 / EXE 所在）
    # ================================
    script_dir = get_app_dir()

    # ================================
    # 工程根目录
    # Version/
    # └── ..
    # ================================
    project_root = script_dir.parent

    # ================================
    # 工具链
    # ================================
    armcc = KEIL_PATH / "ARM" / "ARMCC" / "bin" / "armcc.exe"
    armar = KEIL_PATH / "ARM" / "ARMCC" / "bin" / "armar.exe"

    must_exist(armcc, "armcc")
    must_exist(armar, "armar")

    # ================================
    # 输出路径
    # ================================
    obj_dir = project_root / "output"
    lib_path = script_dir / "version.lib"

    obj_dir.mkdir(parents=True, exist_ok=True)

    # 清理旧文件
    if lib_path.exists():
        lib_path.unlink()

    for obj in obj_dir.glob("*.o"):
        obj.unlink()

    # ================================
    # 编译参数（ARMCC5 对齐 Keil）
    # ================================
    cflags = [
        "--c99",
        "--cpu", "Cortex-M4.fp",
        "-O0",
        "--apcs=interwork",
        "--split_sections",
        "-DGD32F450",
        "-I", str(project_root / "gd32Lib" / "Include"),
        "-I", str(script_dir),
        "-c",
    ]

    # ================================
    # 编译 gd32f4xx_fmc.c
    # ================================
    fmc_src = project_root / "gd32Lib" / "Source" / "gd32f4xx_fmc.c"
    must_exist(fmc_src, "gd32f4xx_fmc.c")

    print("[CC] gd32f4xx_fmc.c")
    run([
        str(armcc),
        *cflags,
        str(fmc_src),
        "-o", str(obj_dir / "gd32f4xx_fmc.o"),
    ])

    # ================================
    # 编译 version.c
    # ================================
    version_c = script_dir / "version.c"
    must_exist(version_c, "version.c")

    print("[CC] version.c")
    run([
        str(armcc),
        *cflags,
        str(version_c),
        "-o", str(obj_dir / "version.o"),
    ])

    # ================================
    # 打包静态库
    # ================================
    print("[AR] version.lib")
    run([
        str(armar),
        "--create",
        str(lib_path),
        str(obj_dir / "gd32f4xx_fmc.o"),
        str(obj_dir / "version.o"),
    ])

    # ================================
    # 可选：删除 version 源文件
    # ================================
    if CLEAN_VERSION_SRC == 1:
        print("[INFO] Cleaning version source files")
        (script_dir / "version.c").unlink(missing_ok=True)
        (script_dir / "version.h").unlink(missing_ok=True)

    print("[SUCCESS] Build finished:", lib_path)

if __name__ == "__main__":
    main()

File: Version/旧方案处理/build_lib.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
)

// ================================
// 构建策略开关
// 0 = 保留 version.c / version.h
// 1 = 生成 lib 后删除
// ================================
const CLEAN_VERSION_SRC = 0

// ================================
// Keil 安装路径
// ================================
const KEIL_PATH = `E:\Keil`

func run(cmd string, args ...string) error {
	// fmt.Printf("[CMD] %s %v\n", cmd, args)
	c := exec.Command(cmd, args...)
	c.Stdout = os.Stdout
	c.Stderr = os.Stderr
	return c.Run()
}

func mustExist(path string, name string) {
	if _, err := os.Stat(path); err != nil {
		fmt.Printf("[ERROR] %s not found: %s\n", name, path)
		os.Exit(1)
	}
}

func main() {
	fmt.Println("[INFO] Build GD32 FMC + Version static library (ARMCC5)")

	// ================================
	// 当前 exe / go 文件所在目录（Version）
	// ================================
	exePath, err := os.Executable()
	if err != nil {
		panic(err)
	}
	scriptDir := filepath.Dir(exePath)

	// ================================
	// 工程根目录
	// ================================
	projectRoot := filepath.Clean(filepath.Join(scriptDir, ".."))

	// ================================
	// 工具链
	// ================================
	armcc := filepath.Join(KEIL_PATH, "ARM", "ARMCC", "bin", "armcc.exe")
	armar := filepath.Join(KEIL_PATH, "ARM", "ARMCC", "bin", "armar.exe")

	mustExist(armcc, "armcc")
	mustExist(armar, "armar")

	// ================================
	// 路径
	// ================================
	objDir := filepath.Join(projectRoot, "output")
	libPath := filepath.Join(scriptDir, "version.lib")

	os.MkdirAll(objDir, 0755)

	// 清理旧文件
	os.Remove(libPath)
	filepath.Walk(objDir, func(p string, info os.FileInfo, err error) error {
		if filepath.Ext(p) == ".o" {
			os.Remove(p)
		}
		return nil
	})

	// ================================
	// 编译参数（ARMCC5 对齐 Keil）
	// ================================
	cflags := []string{
		"--c99",
		"--cpu", "Cortex-M4.fp",
		"-O0",
		"--apcs=interwork",
		"--split_sections",
		"-DGD32F450",
		"-I", filepath.Join(projectRoot, "gd32Lib", "Include"),
		"-I", scriptDir,
		"-c",
	}

	// ================================
	// 编译 gd32f4xx_fmc.c
	// ================================
	fmt.Println("[CC] gd32f4xx_fmc.c")
	err = run(
		armcc,
		append(cflags,
			filepath.Join(projectRoot, "gd32Lib", "Source", "gd32f4xx_fmc.c"),
			"-o", filepath.Join(objDir, "gd32f4xx_fmc.o"),
		)...,
	)
	if err != nil {
		fmt.Println("[ERROR] Compile failed")
		os.Exit(1)
	}

	// ================================
	// 编译 version.c
	// ================================
	fmt.Println("[CC] version.c")
	err = run(
		armcc,
		append(cflags,
			filepath.Join(scriptDir, "version.c"),
			"-o", filepath.Join(objDir, "version.o"),
		)...,
	)
	if err != nil {
		fmt.Println("[ERROR] Compile failed")
		os.Exit(1)
	}

	// ================================
	// 打包静态库
	// ================================
	fmt.Println("[AR] version.lib")
	err = run(
		armar,
		"--create",
		libPath,
		filepath.Join(objDir, "gd32f4xx_fmc.o"),
		filepath.Join(objDir, "version.o"),
	)
	if err != nil {
		fmt.Println("[ERROR] Archive failed")
		os.Exit(1)
	}

	// ================================
	// 可选：删除 version 源文件
	// ================================
	if CLEAN_VERSION_SRC == 1 {
		fmt.Println("[INFO] Cleaning version source files")
		os.Remove(filepath.Join(scriptDir, "version.c"))
		os.Remove(filepath.Join(scriptDir, "version.h"))
	}

	fmt.Println("[SUCCESS] Build finished:", libPath)
}
